### ISSUE TITLE
Mock protocol with associated type constraints.

### ISSUE THREAD

**Issue by grsouza** [MakeAWishFoundation/SwiftyMocky#216]
Hi, awesome framework.

I have a doubt, is it possible to mock a protocol like this one:

```swift
// sourcery: AutoMockable
protocol SuggestionRepository: Repository where Entity == Suggestion {}
```
The Repository protocol definition is:

```swift
protocol Repository {
  associatedtype Entity: EntityType

  func save(entity: Entity) -> Completable
  func save(entities: [Entity]) -> Completable

  func find(
    where predicate: NSPredicate,
    sortedBy sortDescriptors: [NSSortDescriptor]) -> Single<[Entity]>

  func findOne(where predicate: NSPredicate) -> Maybe<Entity>
  func delete(entity: Entity) -> Completable
  func delete(entities: [Entity]) -> Completable
}
```

When a try to mock the `SuggestionRepository`like above, the compiler gives me the following error:
`Reference to invalid associated type 'Entity' of type 'SuggestionRepositoryMock'`.

Thanks.

**Comment by amichnia** [comment]
Hi. I am not able to check it now. But due to some limitations you need to ”annotate” associated types. I don't know if it helps now, but please look at the last section at https://cdn.rawgit.com/MakeAWishFoundation/SwiftyMocky/3.3.4/docs/handling-generics.html

**Comment by amichnia** [comment]
I'm adding this example to sample project, we'll see if it's possible to generate valid mocks for that. @grsouza please stay tuned :)

**Comment by amichnia** [comment]
@grsouza I will have it released this week. So far I came up with this solution:

```swift
//sourcery: AutoMockable
//sourcery: typealias = "Entity = Suggestion"
protocol SuggestionRepository: Repository where Entity == Suggestion {}
```

I added new "annotation" called typealias, I believe it would solve more than this one limitation. That allows you to inject type aliases into generated mocks, making it possible to define what Entity means in a Mock (and so satisfying the constraint).

Unfortunately, there is no automatic detection of such cases yet, but at this point there is little we can do. I will investigate this topic further, maybe recent releases of Sourcery made these things easier.
